client/my-sites/checkout/composite-checkout/lib/get-flow-plan-features.ts

```diff
@@ -33,10 +46,15 @@ export default function getFlowPlanFeatures(
 	}
 
 	const highlightedFeatures = getHighlightedFeatures( flowName, planConstantObj );
-	return getPlanFeaturesObject( featureAccessor() ).map( ( feature ) => {
-		return {
-			...feature,
-			isHighlightedFeature: highlightedFeatures.includes( feature.getSlug() ),
-		};
-	} );
+	const showFreeDomainFeature = ! hasDomainsInCart && ! hasRenewalInCart && nextDomainIsFree;
+	return getPlanFeaturesObject( ( featureAccessor as () => string[] )() )
```

Comment:
I believe this line caused the fatal errors that resulted in [the revert](https://github.com/Automattic/wp-calypso/pull/77317) of this PR.

`featureAccessor` comes from calling `getPlanFeatureAccessor()`, but the return type of that function is very complicated and I think it might be buggy.

Currently, it looks like this:

```ts
	return [
		newsletterFeatures( flowName, plan ),
		linkInBioFeatures( flowName, plan ),
		hostingFeatures( flowName, plan ),
		signupFlowDefaultFeatures( flowName, plan, isInVerticalScrollingPlansExperiment ),
	].find( ( accessor ) => {
		return accessor instanceof Function;
	} );
```

The way it's used here the code expects the return value to be `)() => string[])|undefined`, but there's two problems with that assumption which you can see more easily if we add a return type to the function itself:

```ts
export const getPlanFeatureAccessor = ( {
	flowName = '',
	isInVerticalScrollingPlansExperiment = false,
	plan,
}: {
	flowName?: string;
	isInVerticalScrollingPlansExperiment?: boolean;
	plan: IncompleteWPcomPlan;
} ): undefined | ( () => string[] ) => {
	return [
		newsletterFeatures( flowName, plan ),
		linkInBioFeatures( flowName, plan ),
		hostingFeatures( flowName, plan ),
		signupFlowDefaultFeatures( flowName, plan, isInVerticalScrollingPlansExperiment ),
	].find( ( accessor ) => {
		return accessor instanceof Function;
	} );
};
```

1. The first is more of a TS issue. Each of the sub-functions in `getPlanFeatureAccessor` can return `false`, so in order to properly operate on the ones that don't, we need to convince TS that the falsy ones have been removed. You can do that with `isValueTruthy` like this:

```ts
	return [
		newsletterFeatures( flowName, plan ),
		linkInBioFeatures( flowName, plan ),
		hostingFeatures( flowName, plan ),
		signupFlowDefaultFeatures( flowName, plan, isInVerticalScrollingPlansExperiment ),
	]
		.filter( isValueTruthy )
		.find( ( accessor ) => {
			return accessor instanceof Function;
		} );
```

2. Unlike the others, the `hostingFeatures` sub-function does not return `() => string[]`. It returns `(term?: string) => () => string[]`. This makes `getPlanFeatureAccessor` difficult to use since it sometimes returns a function that returns strings and sometimes returns a function that returns another function. If you remove `hostingFeatures`, you'll see that the types now work correctly and you don't need `as` here:

```ts
export const getPlanFeatureAccessor = ( {
	flowName = '',
	isInVerticalScrollingPlansExperiment = false,
	plan,
}: {
	flowName?: string;
	isInVerticalScrollingPlansExperiment?: boolean;
	plan: IncompleteWPcomPlan;
} ): undefined | ( () => string[] ) => {
	return [
		newsletterFeatures( flowName, plan ),
		linkInBioFeatures( flowName, plan ),
		signupFlowDefaultFeatures( flowName, plan, isInVerticalScrollingPlansExperiment ),
	]
		.filter( isValueTruthy )
		.find( ( accessor ) => {
			return accessor instanceof Function;
		} );
};
```

You could try to guard against that return value, but I don't know what the best solution is because I don't know how `hostingFeatures` is intended to be used. 

My guess is that when `getPlanFeaturesObject` was passed a function instead of an array of strings, the fatal error `planFeaturesList.map is not a function` was thrown.

Any time you find yourself reaching for the `as` keyword to fix a TypeScript error, please consider that there might be something else going on!

Reply:
To build on this, the `hostingFeatures` function used to call `plan.getHostingSignupFeatures` and now it does not. That explains the transition from `() => string[]` to `(term?: string) => () => string[]`

Reply:
If I remember correctly, when I reviewed it `getHostingSignupFeatures` had the `(term) => () => string` wrapper removed, so removing the function call made sense at the time. At some point, it was moved back to its previous syntax, but restoring the call with `term` wasn't.

Reply:
Thanks for the feedback, @sirbrillig & @pottedmeat. I have tried in vain to reproduce the issue by buying monthly/annual plans, adding custom domains, buying add-ons, buying just a domain, and using the tailored flows. The checkout summary page always loads. Were any of you able to reproduce the issue?